src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,257 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+            return BindPattern(node.Pattern, Compilation.GetSpecialType(SpecialType.System_Int32), ExternalScope, permitDesignations, hasErrors, diagnostics);
+        }
+
+        private void LookupLengthOrCountProperty(
+            SyntaxNode node, TypeSymbol inputType, out PropertySymbol? lengthProperty,
+            ref bool hasErrors, BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.IsKind(SyntaxKind.ListPatternClause) || node.IsKind(SyntaxKind.LengthPatternClause));
+
+            var lookupResult = LookupResult.GetInstance();
+            var useSiteInfo = CompoundUseSiteInfo<AssemblySymbol>.Discarded;
+            if (!TryLookupLengthOrCount(inputType, lookupResult, out lengthProperty, ref useSiteInfo))
+            {
+                if (!inputType.IsErrorType() && !hasErrors)
+                {
+                    var errorCode = node.IsKind(SyntaxKind.ListPatternClause)
+                        ? ErrorCode.ERR_UnsupportedTypeForListPattern
+                        : ErrorCode.ERR_UnsupportedTypeForLengthPattern;
+                    Error(diagnostics, errorCode, node, inputType.ToDisplayString());
+                }
+                hasErrors = true;
+            }
+            lookupResult.Free();
+        }
+
+        private BoundListPatternClause BindListPatternClause(
+            PropertyPatternClauseSyntax node,
+            TypeSymbol inputType,
+            bool permitDesignations,
+            ref PropertySymbol? lengthProperty,
+            ref bool hasErrors,
+            BindingDiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.IsKind(SyntaxKind.ListPatternClause));
+
+            if (lengthProperty is null)
+            {
+                LookupLengthOrCountProperty(node, inputType, out lengthProperty, ref hasErrors, diagnostics);
+            }
+
+            TypeSymbol elementType;
+            BoundIndexerAccess? indexerAccess = null;
+            PropertySymbol? indexerSymbol = null;
+            if (inputType.IsSZArray())
+            {
+                elementType = ((ArrayTypeSymbol)inputType).ElementType;
+            }
+            else if (TryFindIndexerOrIndexerPattern(node, inputType, argIsIndex: true, out indexerAccess, out Symbol? patternSymbol, diagnostics))
+            {
+                if (patternSymbol is PropertySymbol indexer)
+                {
+                    indexerSymbol = indexer;
+                    elementType = indexer.Type;
+                }
+                else
+                {
+                    Debug.Assert(indexerAccess is not null);
+                    elementType = indexerAccess.Type;
+                }
+            }
+            else
+            {
+                // PROTOTYPE(list-patterns) Cover any other type that we intend to cover, multi-dimensional arrays?
+                if (!hasErrors && !inputType.IsErrorType())
```

Comment:
Consider removing the `hasErrors` check here. If there's a problem with the counting and the indexing, then the user likely needs to address both.

Reply:
That would result in duplicate errors, because we may have produced this message earlier in length lookup.

See `ListPattern_MissingMembers` to ensure it matches your expectation.

Reply:
Here's the kind of scenario where I think it would be beneficial to also show the error on indexer issue for list pattern (we already report issue with length and slice). I've also noticed this in a test with static Length/indexer/Slice from IL.

```
        [Fact]
        public void ListPattern_Dynamic()
        {
            var source = @"
class C
{
    void M(dynamic d)
    {
        _ = d is [_] { _, .. };
    }
}
";
            var compilation = CreateCompilation(source, parseOptions: TestOptions.RegularWithListPatterns);
            compilation.VerifyEmitDiagnostics(
                // (6,18): error CS9202: Length patterns may not be used for a value of type 'dynamic'.
                //         _ = d is [_] { _, .. };
                Diagnostic(ErrorCode.ERR_UnsupportedTypeForLengthPattern, "[_]").WithArguments("dynamic").WithLocation(6, 18),
                // (6,27): error CS9201: Slice patterns may not be used for a value of type 'dynamic'.
                //         _ = d is [_] { _, .. };
                Diagnostic(ErrorCode.ERR_UnsupportedTypeForSlicePattern, "..").WithArguments("dynamic").WithLocation(6, 27)
                );
        }
```

Reply:
for the record, this is gonna change. `..` won't give an unsupported type error.

I'll make sure we produce all three for `[_] { _, .. var p }` .

Reply:
Makes sense